### DISCUSSION TITLE
Sonoff Zbbridge not show battery status for tuya zb devices

### DISCUSSION THREAD

**Discussion by midnight1979** [arendst/Tasmota#16115] [Q&A] [answered]
### PROBLEM DESCRIPTION
I have tuya smoke sensor and tuya temperature sensors, temperature is showing and it is OK, but I can't see battery charge status - I think that this attribute (battery charge) is in specific cluster but don't know how to get it.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ ] Tasmota binary firmware version number used: _____
  - [x] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by s-hadinger** [comment, marked as answer]
Tuya zigbee is a giant mess and they use non standard attributes that differ from device to device. 

I can look in Z2M if they have figured out the right attribute. 

What are the devices? Modelid and manufacturer

**Reply by MattWestb** [reply]
Some tuya Zigbee and DP/MCU device is not working normal and need "tuya magic spell" being casted (some at paring other its working anytime) for changing its configuration.
Most famous is the old dimmer switch TS004F that can changing mode from light (command) controller to "tuya scene" (clicks as the TS0044 switch) but is not initiating the 3 new endpoints if not getting the spell casted after the device have getting the network key and before the device is requesting updating the link key = the device is in "init mode".

Its looks like all temperature / humidity sensors with display is needing it (I dont have one so i cant guarantee it) for requesting the time and also reporting attributes OK.

**Comment by midnight1979** [comment]
> Tuya zigbee is a giant mess and they use non standard attributes that differ from device to device. 
> 
> I can look in Z2M if they have figured out the right attribute. 
> 
> What are the devices? Modelid and manufacturer 

"ModelId":"TS0201","Manufacturer":"_TZ3000_bguser20" it is temperature sensor

And
"ModelId":"TS0601","Manufacturer":"_TZE200_ntcy3xu1" - smoke sensor

**Reply by s-hadinger** [reply]
From Z2M, `_TZ3000_bguser20` reports battery level using the standard Zigbee ZCL attributes. They should pop up one in a while.

You could try to force attribute reporting on the values `BatteryVoltage` and `BatteryPercentage`. Beware that this may cause some battery drain on some devices

**Reply by midnight1979** [reply]
> From Z2M, `_TZ3000_bguser20` reports battery level using the standard Zigbee ZCL attributes. They should pop up one in a while.
> 
> You could try to force attribute reporting on the values `BatteryVoltage` and `BatteryPercentage`. Beware that this may cause some battery drain on some devices

But in tasmota on zbbridge there is no any battery icon and I never seen that these attributes reporting from device in web log.

**Reply by s-hadinger** [reply]
One thing you could try as a quick test is create a rule to read those both attributes when you receive the value of a sensor. This should work since the device is still awake. Just be careful to avoid an infinite loop

**Reply by s-hadinger** [reply]
If this works, I can try to bake in similar logic by default

**Reply by midnight1979** [reply]
It is OK with temp sensor, it was work for 1 day and had reported about battery charge.

**Reply by midnight1979** [reply]
I have motion sensor with luminosity sensor in one _TZE200_3towulqd works about half of month and not report battery charge.

**Reply by s-hadinger** [reply]
Do you have a log entry of the device reporting any attribute?

**Comment by s-hadinger** [comment]
You will be able to try the new #16148 feature in a couple of hours. The new logic is to explicitly ask for battery level whenever a sensor data is reported (hence we know the device is awake), no more than once per day. To prevent from excessive traffic, the battery level is asked max once per hour if nothing was reported during the last attempt.

**Comment by midnight1979** [comment]
> You will be able to try the new #16148 feature in a couple of hours. The new logic is to explicitly ask for battery level whenever a sensor data is reported (hence we know the device is awake), no more than once per day. To prevent from excessive traffic, the battery level is asked max once per hour if nothing was reported during the last attempt.

OK, I'll try! Thanks!

**Comment by midnight1979** [comment]
i have noticed that "ModelId":"TS0601","Manufacturer":"_TZE200_3towulqd"
report several times per day payload with EF00/0204:100 i think it is a battery percentage, but tasmota don't understand this data.

**Reply by s-hadinger** [reply]
It is not mentioned in Zigbee2MQTT, but it looks like it.

For this, you can now include a special file `tuya_luminance_ts0601.zb` with content:

```
#Z2Tv1
# Luminance motion sensor
:TS0601, _TZE200_3towulqd
EF00/0204=0001/0021,mul:2           # match to BatteryPercentage
```

It requires a very recent version of Tasmota, and check in the logs that the file was sucessfully loaded at the end of the Zigbee initialization process.

**Reply by midnight1979** [reply]
I need an advice how to include this file. Where I must do this? I must to compile tasmota from sources?

**Reply by s-hadinger** [reply]
Just drop this file in the file-system of the Zigbee coordinator. Here is a real file to ease the process. I need to put it in a zip file or Github refuses it.

[tuya_luminance_ts0601.zb.zip](https://github.com/arendst/Tasmota/files/9416304/tuya_luminance_ts0601.zb.zip)

**Reply by midnight1979** [reply]
Can I rename this file? because it is for motion sensor

**Reply by sfromis** [reply]
Yes

**Reply by s-hadinger** [reply]
As long as the suffix is `.zb`

**Reply by midnight1979** [reply]
one question - where i can read about syntax of .zb files? what does meen "mul:2"??

**Reply by midnight1979** [reply]
and also this sensor reporting illuminance on EF00/020C
can i add into the .zb file string of kind: EF00/020C=0400/0000 ? i have read that supported cluster for tuya luminance is 0400 and attr ID is 0000

**Reply by sfromis** [reply]
You can read about this very new feature here: https://github.com/arendst/Tasmota/pull/16252

**Reply by midnight1979** [reply]
i tried to upload this file but i'm not see anything in console that this files was processed
[Tasmota 12.0.2]

**Reply by sfromis** [reply]
You did unzip it? And is the file available in the file system, including when you click the edit button?
Better upgrade to the very latest dev binary from http://ota.tasmota.com/tasmota/

**Reply by midnight1979** [reply]
upgraded to 12.1.0 seems that not working, ofcourse file unziped

**Reply by midnight1979** [reply]
upgraded to dev release 12.1.0.1 - AND IT IS WORKING!
checked on lluminanse - works great, now waiting for battery report, i think it will work too!
Thanks a lot!

**Reply by s-hadinger** [reply]
It's not in 12.1.0. You need to get the latest from development branch 12.1.0.1.

`mul:2` means that the value needs to be multiplied by `2`. The standard `BatteryPercentage` has a .5 % resolution ranging from 0 to 200.

`EF00/020C=0400/0000` would indeed convert any value received as an illuminance value.

**Reply by midnight1979** [reply]
> It's not in 12.1.0. You need to get the latest from development branch 12.1.0.1.
> 
> `mul:2` means that the value needs to be multiplied by `2`. The standard `BatteryPercentage` has a .5 % resolution ranging from 0 to 200.
but sensor reporting 100 (not 200), it is right to multiply by 2 ?
> 
> `EF00/020C=0400/0000` would indeed convert any value received as an illuminance value.
this convertion i can make in node-red

**Reply by midnight1979** [reply]
and the last question about this sensor:
it report motion event as: "EF00/0401":0 - it mean that "motion is detected/occupied" and if it report 1 on this attr - that mean "no motion/not occupied"...

it is posible to invert this attr value with .zb file rules????

**Reply by s-hadinger** [reply]
Oh my.  Really?!? They can go very far into non standard. 

Currently it's not possible to invert a boolean.

**Reply by s-hadinger** [reply]
I need to add this option.

**Reply by s-hadinger** [reply]
Sorry thinking again. Type 04 means 32 bits integer. We need to check that there are no other possible values than 1 and 0

**Reply by midnight1979** [reply]
i think that the enduser like me must decide how to convert attrs like this and must understand what he doing (if i'm sure that possible attr value can be 1 or 0 only - i'll be able to make this convertion into boolean with invertion or not).
Some tuya zigbee devices ofcourse can report different values to this cluster, but rules for specific device in .zb file will be written with these specific features.

**Reply by midnight1979** [reply]
> Oh my.  Really?!? They can go very far into non standard. 
> 
The same situation with the smoke detector sensor (0 - alarm, 1 - normal)
Battery status on another cluster/attr reports like 2 - OK battery, 0 - battery is low )))

**Reply by MattWestb** [reply]
All TS004F devices (dimmer switches with "tuya scene" / events) can being added as one light to one Zigbee group but not bound to one Zgbee group like all controller device shall doing.
The worse is that tuya **is** member of CSA but is only certificating device for some brands like LIDL and all other is not certificated = not Zigbee standard.

**Comment by s-hadinger** [comment]
I have some good news. With the PR #16340 (soon to be merged), you can do such transformation.

`EF00/0401=0406/0000,mul:-1,offset:1`

**Reply by s-hadinger** [reply]
Last minute change, `offset` is too ambiguous, I changed to:

`EF00/0401=0406/0000,mul:-1,add:1`

**Comment by midnight1979** [comment]
It's great, I'll check it as soon as possible, thanks a lot!

**Reply by midnight1979** [reply]
Works great, thank a lot!
